### PR TITLE
Load AES keys stored in boot9.bin

### PR DESCRIPTION
**Pull request by B3n30**
This will load the AES keys that are stored in the boot9.bin. Keys stored in the aes_keys.txt will overwrite any of the keys loaded from the bootrom.

The bootrom boot9.bin can be dumped with Godmode9:
- go to drive M: Memory Virtual
- navigate to boot9.bin, press A
- Select Copy to O:/gm9out
=> This will give you boot9.bin in gm9out on your SD card

The boot9.bin needs to be placed in `sysdata` in the user directory.

Note that there are still keys that aren't stored in the bootrom so to have complete decryption support the following keys need to be provided inside aes_keys.txt:
```
generator
slot0x18KeyX
slot0x1BKeyX
slot0x25KeyX
slot0x31KeyY
common0
common1 
```

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4335)

### CODE REVIEW

File: src/core/hw/aes/key.cpp [hunk 1]
```diff
@@ -70,6 +76,79 @@ AESKey HexToKey(const std::string& hex) {
...
+         {'N', 0x24, false}, {'N', 0x25, true},  {'N', 0x26, true},  {'N', 0x27, true},
+         {'N', 0x28, true},  {'N', 0x29, false}, {'N', 0x2A, false}, {'N', 0x2B, false},
+         {'N', 0x2C, false}, {'N', 0x2D, true},  {'N', 0x2E, true},  {'N', 0x2F, true},
+         {'N', 0x30, false}, {'N', 0x31, true},  {'N', 0x32, true},  {'N', 0x33, true},
+         {'N', 0x34, false}, {'N', 0x35, true},  {'N', 0x36, true},  {'N', 0x37, true},
+         {'N', 0x38, false}, {'N', 0x39, true},  {'N', 0x3A, true},  {'N', 0x3B, true},
+         {'N', 0x3C, true},  {'N', 0x3D, false}, {'N', 0x3E, false}, {'N', 0x3F, false}}};
+
+    // Some of the normal keys retreived here aren't used/valid and needs to get overwritten
+    // by other methodes.
```
Comment (FearlessTobi) [inline src/core/hw/aes/key.cpp]: Some of the normal keys retreived here aren't used/valid and **need** to get overwritten
by other **methods**.

File: src/core/hw/aes/key.cpp [hunk 2]
```diff
@@ -70,6 +77,76 @@ AESKey HexToKey(const std::string& hex) {
...
+    if (length != 65536) {
+        LOG_ERROR(HW_AES, "Bootrom9 size is wrong: {}", length);
+        return;
+    }
+
+    constexpr std::size_t KEY_SECTION_START = 55760;
+    file.Seek(KEY_SECTION_START, 0); // Jump to the key section
+
+    AESKey new_key;
+    for (auto key : keys) {
```
Comment (lioncash) [inline src/core/hw/aes/key.cpp]: This should likely be `const auto&`

File: src/core/hw/aes/key.cpp [hunk 3]
```diff
@@ -70,6 +77,76 @@ AESKey HexToKey(const std::string& hex) {
...
+    // Some of the normal keys retreived here aren't used/valid and need to get overwritten
+    // by other methods.
+
+    const std::string filepath = FileUtil::GetUserPath(FileUtil::UserPath::SysDataDir) + BOOTROM9;
+    auto file = FileUtil::IOFile(filepath, "rb");
+    if (!file) {
+        return;
+    }
+
+    std::size_t length = file.GetSize();
```
Comment (lioncash) [inline src/core/hw/aes/key.cpp]: This can be const.

File: src/core/hw/aes/key.cpp [hunk 4]
```diff
@@ -70,6 +77,76 @@ AESKey HexToKey(const std::string& hex) {
...
+        return;
+    }
+
+    constexpr std::size_t KEY_SECTION_START = 55760;
+    file.Seek(KEY_SECTION_START, 0); // Jump to the key section
+
+    AESKey new_key;
+    for (auto key : keys) {
+        if (!key.same_as_before) {
+            file.ReadArray(reinterpret_cast<char*>(new_key.data()), new_key.size());
```
Comment (lioncash) [inline src/core/hw/aes/key.cpp]: This should probably handle the case where the read doesn't complete successfully.

File: src/core/hw/aes/key.cpp [hunk 5]
```diff
@@ -70,6 +77,76 @@ AESKey HexToKey(const std::string& hex) {
     return key;
 }
 
+void LoadBootromKeys() {
+    const std::array<KeyDesc, 80> keys = {
```
Comment (lioncash) [inline src/core/hw/aes/key.cpp]: This can be `constexpr`

File: src/core/hw/aes/key.cpp [hunk 6]
```diff
@@ -70,6 +77,76 @@ AESKey HexToKey(const std::string& hex) {
...
+    }
+
+    std::size_t length = file.GetSize();
+    if (length != 65536) {
+        LOG_ERROR(HW_AES, "Bootrom9 size is wrong: {}", length);
+        return;
+    }
+
+    constexpr std::size_t KEY_SECTION_START = 55760;
+    file.Seek(KEY_SECTION_START, 0); // Jump to the key section
```
Comment (lioncash) [inline src/core/hw/aes/key.cpp]:
```cpp
file.Seek(KEY_SECTION_START, SEEK_SET);
```

There's no guarantee that `SEEK_SET` is always `0` on every platform, only that it's a distinct integer value from `SEEK_CUR` and `SEEK_END`.

File: src/core/hw/aes/key.cpp (hunk 4 above)
Comment (wwylele) [reply src/core/hw/aes/key.cpp]: Also you don't need to do reinterpret_cast here

Comment (B3n30) [reply src/core/hw/aes/key.cpp]: done

File: src/core/hw/aes/key.cpp [hunk 7]
```diff
@@ -70,6 +77,80 @@ AESKey HexToKey(const std::string& hex) {
...
+         {'N', 0x20, false}, {'N', 0x21, true},  {'N', 0x22, true},  {'N', 0x23, true},
+         {'N', 0x24, false}, {'N', 0x25, true},  {'N', 0x26, true},  {'N', 0x27, true},
+         {'N', 0x28, true},  {'N', 0x29, false}, {'N', 0x2A, false}, {'N', 0x2B, false},
+         {'N', 0x2C, false}, {'N', 0x2D, true},  {'N', 0x2E, true},  {'N', 0x2F, true},
+         {'N', 0x30, false}, {'N', 0x31, true},  {'N', 0x32, true},  {'N', 0x33, true},
+         {'N', 0x34, false}, {'N', 0x35, true},  {'N', 0x36, true},  {'N', 0x37, true},
+         {'N', 0x38, false}, {'N', 0x39, true},  {'N', 0x3A, true},  {'N', 0x3B, true},
+         {'N', 0x3C, true},  {'N', 0x3D, false}, {'N', 0x3E, false}, {'N', 0x3F, false}}};
+
+    // Some of the normal keys retreived here aren't used/valid and need to get overwritten
```
Comment (Subv) [inline src/core/hw/aes/key.cpp]: Could you add a better comment please? From reading this i have no idea why the keys obtained from the bootrom wouldn't be valid

File: src/core/hw/aes/key.cpp [hunk 8]
```diff
@@ -20,6 +21,12 @@ namespace {
 
 std::optional<AESKey> generator_constant;
 
+struct KeyDesc {
+    char key_type;
+    std::size_t slot_id;
+    bool same_as_before;
```
Comment (Subv) [inline src/core/hw/aes/key.cpp]: What is this boolean for? please add a comment

**Pull request merged**